rfd/0059-search-based-access-requests.md

```diff
@@ -0,0 +1,250 @@
+---
+authors: Alexander Klizhentas (sasha@goteleport.com), Nic Klaassen (nic@goteleport.com)
+state: draft
+---
+
+# RFD 59 - Search Based Access Requests
+
+## What
+
+This RFD proposes just-in-time search-based access requests which will enable
+users to request access to resources based on a search and/or selection of
+individual resources, rather than having to request one or more roles.
+
+## Why
+
+Teleport users may not know in advance which roles they need or who should grant
+access, just which resource(s) they need access to.
+
+Users would like to avoid having many different custom roles for different
+levels of access.
+
+Requesting access to a "blanket" role violates the principle of least priviledge.
+Most of the time incident responders only need access to 1 or 2 nodes,
+search-based access requests will allow them to request access to only the nodes
+they need.
+
+
+## Details
+
+Users will be able to request access by using search, and roles will be used
+behind the scenes to evaluate who will be responsible for reviewing and granting
+access requests.
+
+Users won’t be asked to think about roles in advance, but will be able to focus
+on finding and requesting access to individual resources.
+
+The existing “role-based” access requests and the new “search-based” access
+requests proposed here are useful for two different scenarios:
+
+- Sometimes it is helpful to request elevated roles in the system, like
+  `db-admin`. For example, Alice may need to run a Ansible playbook on all
+  machines in a given class, or perform some system upgrades or troubleshooting.
+  In this case, role-based access requests work great.
+
+- In other cases, it’s helpful to request access to a subset of resources. In
+  this case, search.
+
+### Example Flow
+
+Alice is a member of the group “splunk”.
+
+She gets an alert that “db-1” is malfunctioning.
+
+Alice needs to locate the "db-1" resource and request access to it.
+
+Alice goes to the “Request access” screen and selects: “select access to resources”
+option that is available alongside the “select access to roles” option.
+
+Alice uses standard (fuzzy search) or advanced search (labels) to enter the
+resource name.
+
+She finds a single resource (e.g. "db-1") or a group of resources by that name and clicks
+"request access".
+
+An access request is created. It is sent to the "db-admins" team members via slack,
+email or other configured channel based on the labels on the requested resources
+(we will add label-based routing to access request plugins).
```

Comment:
I'd be interested in seeing a configuration example for the proposed routing of access requests to the request plugins.

Reply:
I added an example of the new configuration section here 812cc46